File: python/sparklabs/batch/sales/rdd/customer_orders.py
from sparklabs.batch.sales.rdd import utils


def run() -> None:
    rows = (
        utils.customers.map(lambda customer: (customer["customer_id"], customer["customer_id"]))
        .join(utils.orders.map(lambda order: (order["customer_id"], order["customer_id"])))
        .map(lambda item: (item[0], 1))
        .reduceByKey(lambda left, _: left)
        .sortBy(lambda item: item[1], ascending=False)
        .collect()
    )
    for row in rows:
        print(row)


if __name__ == "__main__":
    run()

File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/rdd/CustomerOrders.scala
package io.devorbit.sparklabs.batch.sales.rdd

object CustomerOrders extends App {
  val customerIds = RddUtils.customers.map(customer => (customer.customer_id, customer.customer_id))
  val orderIds = RddUtils.orders.map(order => (order.customer_id, order.customer_id))

  customerIds.join(orderIds)
    .map { case (customerId, _) => (customerId, 1L) }
    .reduceByKey((left, _) => left)
    .sortBy(_._2, ascending = false)
    .collect()
    .foreach(println)
}
